src/transform/src/equivalence_propagation.rs

```diff
@@ -0,0 +1,433 @@
+                        if index != other {
+                            push_equivalences
+                                .classes
+                                .extend(input_equivs.classes.clone());
+                        }
+                    }
+                    push_equivalences.project(columns..(columns + child_arity));
+                    self.apply(expr, child, push_equivalences, get_equivalences);
+
+                    columns += child_arity;
+                }
+
+                equivalences.clone_from(&join_equivalences.classes);
+            }
+            MirRelationExpr::Reduce {
+                input,
+                group_key,
+                aggregates: _,
+                ..
+            } => {
+                // TODO: MIN, MAX, ANY, ALL aggregates pass through all certain properties of their columns.
+                // This may involve projection and permutation, to reposition the information appropriately.
+                // TODO: Non-null constraints likely push down into the support of the aggregate expressions.
+
+                // To transform `outer_equivalences` to one about `input`, we will "pretend" to pre-pend all of
+                // the input columns, introduce equivalences about the evaluation of `group_key` on them
+                // and the key columns themselves, and then project onto these "input" columns.
+                let input_arity = *derived
+                    .last_child()
+                    .value::<Arity>()
+                    .expect("Arity required");
+                let output_arity = *derived.value::<Arity>().expect("Arity required");
+
+                // Permute `outer_equivalences` to reference columns `input_arity` later.
+                let permutation = (input_arity..(input_arity + output_arity)).collect::<Vec<_>>();
+                outer_equivalences.permute(&permutation[..]);
+                for (index, group) in group_key.iter().enumerate() {
+                    outer_equivalences.classes.push(vec![
+                        MirScalarExpr::Column(input_arity + index),
+                        group.clone(),
+                    ]);
+                }
+                outer_equivalences.project(0..input_arity);
+                self.apply(
+                    input,
+                    derived.last_child(),
+                    outer_equivalences,
+                    get_equivalences,
+                );
+            }
+            MirRelationExpr::TopK {
+                input,
+                group_key,
+                limit,
+                ..
+            } => {
+                if let Some(expr) = limit {
+                    let input_equivalences = derived
+                        .last_child()
+                        .value::<Equivalences>()
+                        .expect("Equivalences required");
+                    input_equivalences.reduce_expr(expr);
+                }
+                outer_equivalences.project(0..group_key.len());
```

Comment:
I don't know why this line is here, but now I'm scared to remove it.

Reply:
I have discovered why it is there (it is correct!). Comment added above that now explains it.